### PR TITLE
Task push mode using reactivex.

### PR DESCRIPTION
**Pull request by ibuenros**
This PR changes the way Tasks are executed to a push model using Reactivex (http://reactivex.io/).

Main changes:
* Task no longer handles records, instead it handles a record stream containing a Reactivex `Flowable` along with metadata.
* Constructs modified to handle the stream instead of individual records:
  * Extractor: Main point of interaction is `recordStream` method. By default, create a stream that repeatedly calls `readRecord`.
  * Converter / Quality checker: Main point of interaction is `processStream` method. By default, map records using `convertRecord` and `executePolicies`.
  * Writer: minimal changes, mostly around `RecordEnvelope` explained later.
* Forks work a little different. A `Forker` takes a `ForkOperator` and splits the record stream into multiple record streams. Then, we call `Fork.consumeRecordStream` for each such mirrored stream.
* All records are wrapped in `RecordEnvelope` (introduced previously). This is a first step to more closely unify batch and streaming. There's nothing special about `RecordEnvelope`, everything works transparently, except metrics get a little tricky.

### CODE REVIEW

**Comment by ibuenros** [comment]
@ydai1124 can you review? @shirshanka you might be interested.

File: gobblin-api/src/main/java/gobblin/fork/Forker.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    List<Boolean> forkedSchemas = forkOperator.forkSchema(workUnitState, inputStream.getSchema());
+
+    Preconditions.checkState(forkedSchemas.size() == branches, String
+        .format("Number of forked schemas [%d] is not equal to number of branches [%d]", forkedSchemas.size(),
+            branches));
+
+    Flowable<RecordWithForkMap<D>> forkedStream = inputStream.getRecordStream().map(r ->
+        new RecordWithForkMap<>(r, forkOperator.forkDataRecord(workUnitState, r.getRecord())));
+
+    ConnectableFlowable<RecordWithForkMap<D>> connectableFlowable = forkedStream.publish(100);
```
Comment (ydai1124) [inline gobblin-api/src/main/java/gobblin/fork/Forker.java]: Can you make the buffer size configurable instead of hard coded `100`? A single record size can be very large.

File: gobblin-api/src/main/java/gobblin/records/RecordStreamConsumer.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ * An object that a {@link RecordStreamWithMetadata}.
+ * @param <S> schema type
+ * @param <D> data type
+ */
+public interface RecordStreamConsumer<S, D> {
+
+  /**
+   * Subscribe to the input {@link RecordStreamWithMetadata}.
+   */
+  void consumeRecordStream(RecordStreamWithMetadata<D, S> stream) throws Exception;
```
Comment (ydai1124) [inline gobblin-api/src/main/java/gobblin/records/RecordStreamConsumer.java]: Can you create a more specified exception class?

File: gobblin-api/src/main/java/gobblin/records/RecordStreamProcessor.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * @param <SO> output schema type
+ * @param <DI> input data type
+ * @param <DO> output data type
+ */
+public interface RecordStreamProcessor<SI, SO, DI, DO> {
+
+  /**
+   * Return a {@link RecordStreamWithMetadata} with the appropriate modifications.
+   */
+  RecordStreamWithMetadata<DO, SO> processStream(RecordStreamWithMetadata<DI, SI> inputStream, WorkUnitState state)
```
Comment (ydai1124) [inline gobblin-api/src/main/java/gobblin/records/RecordStreamProcessor.java]: ditto.

File: gobblin-api/src/main/java/gobblin/records/RecordStreamWithMetadata.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import lombok.Data;
+
+
+/**
+ * A stream of records along with metadata (e.g. Schema).
+ * @param <D> type of record.
+ * @param <S> type of schema.
+ */
+@Data
+public class RecordStreamWithMetadata<D, S> {
```
Comment (ydai1124) [inline gobblin-api/src/main/java/gobblin/records/RecordStreamWithMetadata.java]: Are there other metadata besides `schema`? Maybe `RecordStreamWithSchema` is more descriptive?

File: gobblin-runtime/src/main/java/gobblin/runtime/fork/Fork.java [hunk 5]
```diff
@@ -440,17 +356,26 @@ protected void cleanup() {
...
+      }
 
-    DataWriter<Object> writer = new PartitionedDataWriter<>(builder, this.taskContext.getTaskState());
-    logger.info("Wrapping writer " + writer);
-    return new DataWriterWrapperBuilder<>(writer, this.taskState).build();
+      DataWriter<Object> writer = new PartitionedDataWriter<>(builder, this.taskContext.getTaskState());
+      logger.info("Wrapping writer " + writer);
+      return new DataWriterWrapperBuilder<>(writer, this.taskState).build();
+    } catch (Throwable t) {
+      logger.error("blah", t);
```
Comment (ydai1124) [inline gobblin-runtime/src/main/java/gobblin/runtime/fork/Fork.java]: Can you replace `blah` with some more descriptive blah blah blah? :)

File: gobblin-runtime/src/main/java/gobblin/runtime/fork/Fork.java [hunk 6]
```diff
@@ -210,51 +222,6 @@ public State getFinalState() {
...
-   * it should not expect more incoming data records.
-   *
-   * <p>
-   *   This method is used by the {@link Task} that creates this {@link Fork}.
-   * </p>
-   */
-  public void markParentTaskDone() {
-    this.parentTaskDone = true;
-    try {
-      this.putRecord(SHUTDOWN_RECORD);
```
Comment (ydai1124) [inline gobblin-runtime/src/main/java/gobblin/runtime/fork/Fork.java]: The definition of constant `SHUTDOWN_RECORD` can be removed?

**Comment by ydai1124** [comment]
There are also build failures.

File: gobblin-api/src/main/java/gobblin/converter/Converter.java [hunk 7]
```diff
@@ -100,4 +105,20 @@ public abstract SO convertSchema(SI inputSchema, WorkUnitState workUnit)
...
+  public RecordStreamWithMetadata<DO, SO> processStream(RecordStreamWithMetadata<DI, SI> inputStream,
+      WorkUnitState workUnitState) throws SchemaConversionException {
+    init(workUnitState);
+    SO outputSchema = convertSchema(inputStream.getSchema(), workUnitState);
+    Flowable<RecordEnvelope<DO>> outputStream =
+        inputStream.getRecordStream()
+            .flatMap(in -> Flowable.fromIterable(convertRecord(outputSchema, in.getRecord(), workUnitState))
+            .map(in::withRecord));
+    outputStream = outputStream.doOnComplete(this::close);
+    return inputStream.withRecordStream(outputStream, outputSchema);
```
Comment (ibuenros) [inline gobblin-api/src/main/java/gobblin/converter/Converter.java]: No, because it preserver all of the metadata in the stream (if present).

File: gobblin-api/src/main/java/gobblin/records/RecordStreamWithMetadata.java (hunk 4 above)
Comment (ibuenros) [inline gobblin-api/src/main/java/gobblin/records/RecordStreamWithMetadata.java]: Not for now, but later we may want to add more stuff to it.

**Comment by shirshanka** [comment]
@ibuenros : High level question is how are you planning to mitigate risk with this change to the core pipeline?

**Comment by ibuenros** [comment]
@shirshanka the plan was to do it through testing, but I understand even that is risky. An alternative is to leave the old pipeline and put a configuration switch, but I don't know how ugly that may get

**Comment by shirshanka** [comment]
@ibuenros : I'd like to think through how we might do that... maybe a new Task impl ?

File: gobblin-api/src/main/java/gobblin/fork/CopyHelper.java [hunk 8]
```diff
@@ -56,37 +56,26 @@ private static boolean isImmutableType(Object thing) {
...
   /**
    * Copy this object if needed.
-   * @param thing: this object that needs to be copied
-   * @param instance: the number of instances of this object already copied
+   * @param thing : this object that needs to be copied
    * @return: a possibly copied instance
    * @throws CopyNotSupportedException if thing needs to be copied but cannot be
    */
-  public static Object copy(Object thing, int instance) throws CopyNotSupportedException {
-    // First copy is special, we only copy if the thing is an instance of Copyable
```
Comment (shirshanka) [inline gobblin-api/src/main/java/gobblin/fork/CopyHelper.java]: Why did you move the check out from here to the caller?

File: gobblin-api/src/main/java/gobblin/source/extractor/Extractor.java [hunk 9]
```diff
@@ -74,4 +83,38 @@
...
+  default RecordEnvelope<D> readRecordEnvelope() throws DataRecordException, IOException {
+    D record = readRecord(null);
+    return record == null ? null : new RecordEnvelope<>(record);
+  }
+
+  /**
+   * @param shutdownRequest an {@link AtomicBoolean} that becomes true when a shutdown has been requested.
+   * @return a {@link Flowable} with the records from this source. Note the flowable should honor downstream backpressure.
+   */
+  default RecordStreamWithMetadata<D, S> recordStream(AtomicBoolean shutdownRequest) throws IOException {
```
Comment (shirshanka) [inline gobblin-api/src/main/java/gobblin/source/extractor/Extractor.java]: What is the expected behavior when shutdownRequest is set to true? 

pros and cons of doing it this way versus having an explicit shutdown method?

File: gobblin-api/src/main/java/gobblin/records/RecordStreamWithMetadata.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+  private final Flowable<RecordEnvelope<D>> recordStream;
+  private final S schema;
```
Comment (eogren) [inline gobblin-api/src/main/java/gobblin/records/RecordStreamWithMetadata.java]: Hmm - if we are consuming from a Kafka topic in streaming mode and the schema is evolved, would we be able to pick up the change with this approach? Should the Schema actually be part of the envelope?

File: gobblin-api/src/main/java/gobblin/fork/CopyHelper.java (hunk 8 above)
Comment (ibuenros) [inline gobblin-api/src/main/java/gobblin/fork/CopyHelper.java]: Because the caller doesn't need to keep track of the instance number, just whether the object needs to be copied or not. For mutable, trivially copyable objects, like byte array, it is incorrect to return `thing` here, as the first fork may mutate the object before it gets copied into other forks (race condition). 

The only instance that is safe to copy is the last one, but then we have an implicit contract that the instances are passed in order to the Copier, which is not specified anywhere and is hard to enforce.

File: gobblin-api/src/main/java/gobblin/records/RecordStreamWithMetadata.java (hunk 10 above)
Comment (ibuenros) [inline gobblin-api/src/main/java/gobblin/records/RecordStreamWithMetadata.java]: This is mostly replicating the current flow where the schema is only passed along once. We could move the schema to be part of the envelope, but then every record would require looking at the schema. At any rate, this design does not prevent us from putting a dummy schema here and the real schema in the envelope.

What I could see as an issue is that on detection of new schema we may want to call pass the schema through the entire flow again. In that case, even in current Gobblin, I believe we would need to re-instantiate the pipeline, as some converters have side effects on `convertSchema`.

File: gobblin-api/src/main/java/gobblin/source/extractor/Extractor.java (hunk 9 above)
Comment (ibuenros) [inline gobblin-api/src/main/java/gobblin/source/extractor/Extractor.java]: I though about that, but having a `shutdown` method is hard because interfaces cannot hold instance variables, so the shutdown has no way of notifying the stream to stop.

File: gobblin-api/src/main/java/gobblin/fork/CopyHelper.java (hunk 8 above)
Comment (ibuenros) [inline gobblin-api/src/main/java/gobblin/fork/CopyHelper.java]: An alternative is to have the method `Iterable<Object> CopyHelper.createCopies(Object obj, int numCopies)`, so the method can guarantee the copies happen before the object is mutated.

File: gobblin-api/src/main/java/gobblin/fork/Forker.java (hunk 1 above)
Comment (ibuenros) [inline gobblin-api/src/main/java/gobblin/fork/Forker.java]: Fixed

File: gobblin-api/src/main/java/gobblin/fork/Forker.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+
+    List<RecordStreamWithMetadata<D, S>> forkStreams = Lists.newArrayList();
+
+    for(int i = 0; i < forkedSchemas.size(); i++) {
+      if (forkedSchemas.get(i)) {
+        final int idx = i;
+        Flowable<RecordEnvelope<D>> thisStream =
+            connectableFlowable.filter(new ForkFilter<>(idx)).map(RecordWithForkMap::getRecordCopyIfNecessary);
+        forkStreams.add(inputStream.withRecordStream(thisStream,
+            mustCopy(forkedSchemas) ? (S) CopyHelper.copy(inputStream.getSchema()) : inputStream.getSchema()));
```
Comment (ibuenros) [inline gobblin-api/src/main/java/gobblin/fork/Forker.java]: Fixed

File: gobblin-api/src/main/java/gobblin/records/RecordStreamConsumer.java (hunk 2 above)
Comment (ibuenros) [inline gobblin-api/src/main/java/gobblin/records/RecordStreamConsumer.java]: Fixed

File: gobblin-api/src/main/java/gobblin/records/RecordStreamProcessor.java (hunk 3 above)
Comment (ibuenros) [inline gobblin-api/src/main/java/gobblin/records/RecordStreamProcessor.java]: Fixed

File: gobblin-runtime/src/main/java/gobblin/runtime/fork/Fork.java (hunk 6 above)
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/fork/Fork.java]: Restored old fork execution, so this will stay for now.

File: gobblin-runtime/src/main/java/gobblin/runtime/fork/Fork.java (hunk 5 above)
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/fork/Fork.java]: Fixed. Thanks for noticing :)

**Comment by ibuenros** [comment]
@shirshanka restored old task execution (albeit refactored into a separate method), which is the default execution method. However, it does still use `RecordEnvelopes` as that is necessary to start unifying stream and batch.
